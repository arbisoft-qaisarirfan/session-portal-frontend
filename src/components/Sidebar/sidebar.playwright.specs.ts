import { test, expect } from "@playwright/test";

import useLanguage from "@/services/i18n/use-language";

test.describe("Sidebar Component", () => {
  const language = useLanguage();
  test("should render all sidebar items", async ({ page }) => {
    await page.goto(`/${language}/videos`);

    const sidebarItems = await page.locator("text=Sample Item 1, Sample Item 2, Sample Item 3");
    await expect(sidebarItems).toBeVisible();
  });

  test("should highlight item as selected when clicked", async ({ page }) => {
    await page.goto(`/${language}/videos`);

    const item = await page.locator("text=Sample Item 1");
    await item.click();

    const selectedItem = await page.locator("text=Sample Item 1");
    await expect(selectedItem).toHaveClass(/selected/);
  });

  test("should trigger sidebar toggle when an item is clicked", async ({ page }) => {
    const toggleButton = await page.locator("[data-testid='sidebar-toggle-button']");
    await toggleButton.click();

    const sidebar = await page.locator("[data-testid='sidebar-container']");
    await expect(sidebar).not.toBeVisible();
  });

  test("should render image with correct alt text for each sidebar item", async ({ page }) => {
    await page.goto(`/${language}/videos`);

    const image = await page.locator("img[alt='Sample Item 1']");
    await expect(image).toBeVisible();
    await expect(image).toHaveAttribute("alt", "Sample Item 1");
  });

  test("should render sidebar items with correct layout", async ({ page }) => {
    await page.goto(`/${language}/videos`);

    const menuStack = await page.locator("div[role='menu']");
    const items = await menuStack.locator("li");
    await expect(items).toHaveCount(3);
  });
});
