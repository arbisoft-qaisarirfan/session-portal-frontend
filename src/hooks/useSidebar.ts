import { useEventTagsQuery } from "@/redux/events/apiSlice";

const defaultTag = { id: 0, name: "All" };

const useSidebar = () => {
  const { data: tags, isFetching, isLoading, isUninitialized } = useEventTagsQuery();
  const isDataLoading = isFetching || isLoading || isUninitialized;

  const sidebarItems = tags?.length ? [defaultTag, ...tags] : [];

  return { sidebarItems, isDataLoading };
};
export default useSidebar;
