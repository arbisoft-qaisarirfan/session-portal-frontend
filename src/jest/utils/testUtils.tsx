import React, { ReactElement } from "react";

import "@testing-library/jest-dom";
import { render, RenderOptions } from "@testing-library/react";

import ThemeProvider from "@/components/theme/theme-provider";
import { Providers } from "@/redux/store/provider";

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Providers>
        <ThemeProvider>{children}</ThemeProvider>
      </Providers>
    ),
    ...options,
  });

export * from "@testing-library/react";

export { customRender };
