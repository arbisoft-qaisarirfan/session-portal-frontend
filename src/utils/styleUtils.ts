export const shouldForwardProp = (propName: string) => !propName.startsWith("$");
