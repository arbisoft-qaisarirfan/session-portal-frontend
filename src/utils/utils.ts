export function secondsToTime(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);

  const formatNumber = (num: number): string =>
    num.toLocaleString(undefined, {
      useGrouping: false,
      minimumIntegerDigits: 2,
    });

  return `${formatNumber(hours)}:${formatNumber(minutes)}`;
}
