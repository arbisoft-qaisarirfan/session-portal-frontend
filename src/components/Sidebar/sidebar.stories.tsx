import type { Meta, StoryObj } from "@storybook/react";

import Sidebar from "./sidebar";

const meta: Meta<typeof Sidebar> = {
  title: "Components/Sidebar",
  component: Sidebar,
  tags: ["autodocs"],
  argTypes: {
    handleSidebarToggle: {
      action: "sidebarToggled",
      description: "Callback function triggered when a sidebar item is clicked",
    },
  },
};

export default meta;

type Story = StoryObj<typeof Sidebar>;

export const Default: Story = {
  args: {},
};
