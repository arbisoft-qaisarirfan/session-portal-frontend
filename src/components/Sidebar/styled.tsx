import Box from "@mui/material/Box";
import { chipClasses } from "@mui/material/Chip";
import Stack from "@mui/material/Stack";
import { styled, css, alpha } from "@mui/material/styles";
import Typography from "@mui/material/Typography";

import { shouldForwardProp } from "@/utils/styleUtils";

export const SidebarContainer = styled(Box, {
  name: "SidebarContainer",
})(() => {
  return css`
    height: 100%;
    width: 100%;
  `;
});

export const MenuStack = styled(Stack, {
  name: "MenuStack",
})(() => {
  return css`
    flex-direction: column;
    height: 100%;
    gap: 8px;
  `;
});

export const MenuItem = styled(Stack, {
  name: "MenuItem",
  shouldForwardProp,
})<{ $isSelected: boolean }>(({ theme, $isSelected }) => {
  const backgroundColor = alpha(theme.palette.colors.white, 0.2);
  return css`
    background-color: ${$isSelected ? backgroundColor : "transparent"};
    border-radius: 8px;
    color: ${theme.palette.colors.white};
    cursor: pointer;
    flex-direction: row;
    gap: 10px;
    padding: 12px;

    &:hover {
      background-color: ${backgroundColor};
    }
  `;
});

export const Text = styled(Typography, {
  name: "Text",
})(() => {
  return css`
    font-size: 14px;
    line-height: 12px;
    font-weight: 500;
    letter-spacing: 0.4px;
  `;
});

export const TagsContainer = styled(Box, {
  name: "TagsContainer",
})(
  ({ theme }) => css`
    display: flex;
    flex-wrap: wrap;
    gap: 5px;
    padding-top: 30px;

    .${chipClasses.root} {
      border-radius: 6px;
      border: 1px solid ${alpha(theme.palette.common.white, 0.3)};
      cursor: pointer;
      text-transform: capitalize;

      &.${chipClasses.filled} {
        background: ${theme.palette.colors.white};
        color: ${alpha(theme.palette.common.black, 0.7)};
      }

      .${chipClasses.icon} {
        margin: 0;
      }
    }
  `
);
