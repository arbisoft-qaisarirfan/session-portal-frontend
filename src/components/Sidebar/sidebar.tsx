import React from "react";

import { faker } from "@faker-js/faker";
import Box from "@mui/material/Box";
import Chip from "@mui/material/Chip";
import Skeleton from "@mui/material/Skeleton";
import Image from "next/image";
import { useSearchParams } from "next/navigation";

import useNavigation from "@/hooks/useNavigation";
import useSidebar from "@/hooks/useSidebar";

import { MenuItem, MenuStack, SidebarContainer, TagsContainer, Text } from "./styled";

const loadingTags: string[] = Array(10)
  .fill("")
  .map(() => faker.lorem.words(1));

const Sidebar = () => {
  const { navigateTo } = useNavigation();
  const searchParams = useSearchParams();

  const { arePlaylistsLoading, tags, playlists } = useSidebar();

  const tag = searchParams?.get("tag");
  const playlist = searchParams?.get("playlist");

  return (
    <SidebarContainer data-testid="sidebar-container">
      <MenuStack>
        <Box>
          {arePlaylistsLoading ? (
            <Box data-testid="loading">
              {loadingTags?.map((item) => (
                <Box key={item} display="flex" justifyContent="space-between" alignItems={"center"} width="90%" mb={1}>
                  <Skeleton variant="rounded" width="15%" height={20} />
                  <Skeleton width="76%" height={25} />
                </Box>
              ))}
            </Box>
          ) : (
            <>
              <MenuItem $isSelected={!playlist} onClick={() => navigateTo("videos")} data-testid={"sidebar-item-All"}>
                <Image src="/assets/images/sidebar-item-icon.svg" alt="All" width={18} height={12} />
                <Text>All</Text>
              </MenuItem>
              {playlists.map((item) => (
                <MenuItem
                  key={item.id}
                  $isSelected={item.name === playlist}
                  onClick={() => navigateTo("videos", { playlist: item.name })}
                  data-testid={`sidebar-item-${item.name}`}
                >
                  <Image src="/assets/images/sidebar-item-icon.svg" alt={item.name} width={18} height={12} />
                  <Text>{item.name}</Text>
                </MenuItem>
              ))}
            </>
          )}
        </Box>
        <TagsContainer data-testid="sidebar-tags">
          <Chip
            data-testid="sidebar-tags-All"
            onClick={() => navigateTo("videos")}
            variant={!tag ? "filled" : "outlined"}
            label="All"
            size="small"
          />
          {tags.map((item) => (
            <Chip
              data-testid={`sidebar-tags-${item.name}`}
              key={item.id}
              onClick={() => navigateTo("videos", { tag: item.name })}
              label={item.name}
              variant={tag === item.name ? "filled" : "outlined"}
              size="small"
            />
          ))}
        </TagsContainer>
      </MenuStack>
    </SidebarContainer>
  );
};

export default Sidebar;
