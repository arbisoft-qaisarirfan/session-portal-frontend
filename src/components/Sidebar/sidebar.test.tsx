import useSidebar from "@/hooks/useSidebar";
import { customRender, fireEvent, screen, waitFor } from "@/jest/utils/testUtils";

import Sidebar from "./sidebar";

jest.mock("@/hooks/useSidebar", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    sidebarItems: [
      { id: 1, name: "Home" },
      { id: 2, name: "Settings" },
      { id: 3, name: "Profile" },
      { id: 4, name: "Logout" },
    ],
  })),
}));

describe("Sidebar Component", () => {
  test("should renders Sidebar without crashing", () => {
    customRender(<Sidebar />);
    expect(screen.getByTestId("sidebar-container")).toBeInTheDocument();
  });

  test("should renders all sidebar items", () => {
    const items = [
      { id: 1, name: "Home" },
      { id: 2, name: "Settings" },
      { id: 3, name: "Profile" },
      { id: 4, name: "Logout" },
    ];
    customRender(<Sidebar />);
    items.forEach((item) => {
      expect(screen.getByText(item.name)).toBeInTheDocument();
    });
  });

  test("should highlights selected item", () => {
    customRender(<Sidebar />);
    const item = screen.getByText("Home");
    fireEvent.click(item);
    expect(screen.getByTestId("sidebar-item-Home")).toBeInTheDocument();
    expect(item.parentElement).toHaveStyle("background-color: rgba(255, 255, 255, 0.2)");
  });

  test("should calls handleSidebarToggle when an item is clicked", () => {
    const handleSidebarToggle = jest.fn();
    customRender(<Sidebar handleSidebarToggle={handleSidebarToggle} />);
    fireEvent.click(screen.getByText("Home"));
    expect(screen.getByTestId("sidebar-item-Home")).toBeInTheDocument();
    expect(handleSidebarToggle).toHaveBeenCalled();
  });

  test("should renders sidebar item icons", () => {
    customRender(<Sidebar />);
    expect(screen.getAllByRole("img")).toHaveLength(4);
  });

  test("should does not call handleSidebarToggle if it is not provided", () => {
    customRender(<Sidebar />);
    fireEvent.click(screen.getByText("Profile"));
    expect(screen.getByTestId("sidebar-item-Profile")).toBeInTheDocument();
    expect(true).toBe(true);
  });

  test("should menu items should be clickable", () => {
    customRender(<Sidebar />);
    const item = screen.getByText("Logout");
    expect(screen.getByTestId("sidebar-item-Logout")).toBeInTheDocument();
    fireEvent.click(item);
    expect(item).toBeTruthy();
  });

  test("should menu should not change selection when clicking on the same item twice", () => {
    customRender(<Sidebar />);
    expect(screen.getByTestId("sidebar-item-Home")).toBeInTheDocument();
    const item = screen.getByText("Home");
    fireEvent.click(item);
    fireEvent.click(item);
    expect(item.parentElement).toHaveStyle("background-color: rgba(255, 255, 255, 0.2)");
  });

  test("should menu should support multiple clicks on different items", async () => {
    customRender(<Sidebar />);
    const item1 = screen.getByText("Home");
    const item2 = screen.getByText("Settings");

    fireEvent.click(item1);
    fireEvent.click(item2);

    await waitFor(() => {
      expect(item2.parentElement).toHaveStyle("background-color: rgba(255, 255, 255, 0.2)");
    });
  });

  test("should not render menu items if sidebarItems is empty", () => {
    (useSidebar as jest.Mock).mockReturnValue({ sidebarItems: [] });
    customRender(<Sidebar />);
    expect(screen.queryByRole("img")).toBeNull();
  });
});
