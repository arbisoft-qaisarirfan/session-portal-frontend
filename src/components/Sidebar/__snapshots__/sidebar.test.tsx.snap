// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component should snapshot test for Sidebar 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-cd9kc0-SidebarContainer-root"
    data-testid="sidebar-container"
  >
    <div
      class="MuiStack-root css-ujfjh4-MuiStack-root-MenuStack-root"
    >
      <div
        class="MuiStack-root css-1gweisr-MuiStack-root-MenuItem-root"
      >
        <img
          alt="Home"
          data-nimg="1"
          decoding="async"
          height="12"
          loading="lazy"
          src="/assets/images/sidebar-item-icon.svg"
          style="color: transparent;"
          width="18"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-9bjbpq-MuiTypography-root-Text-root"
          data-testid="sidebar-item-Home"
        >
          Home
        </p>
      </div>
      <div
        class="MuiStack-root css-1gweisr-MuiStack-root-MenuItem-root"
      >
        <img
          alt="Settings"
          data-nimg="1"
          decoding="async"
          height="12"
          loading="lazy"
          src="/assets/images/sidebar-item-icon.svg"
          style="color: transparent;"
          width="18"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-9bjbpq-MuiTypography-root-Text-root"
          data-testid="sidebar-item-Settings"
        >
          Settings
        </p>
      </div>
      <div
        class="MuiStack-root css-1gweisr-MuiStack-root-MenuItem-root"
      >
        <img
          alt="Profile"
          data-nimg="1"
          decoding="async"
          height="12"
          loading="lazy"
          src="/assets/images/sidebar-item-icon.svg"
          style="color: transparent;"
          width="18"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-9bjbpq-MuiTypography-root-Text-root"
          data-testid="sidebar-item-Profile"
        >
          Profile
        </p>
      </div>
      <div
        class="MuiStack-root css-1gweisr-MuiStack-root-MenuItem-root"
      >
        <img
          alt="Logout"
          data-nimg="1"
          decoding="async"
          height="12"
          loading="lazy"
          src="/assets/images/sidebar-item-icon.svg"
          style="color: transparent;"
          width="18"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-9bjbpq-MuiTypography-root-Text-root"
          data-testid="sidebar-item-Logout"
        >
          Logout
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
